### DISCUSSION TITLE
MQTT topics

### DISCUSSION THREAD

**Discussion by jjvelar** [greghesp/ha-bambulab#480] [Q&A]
Hello,

I really like this integration, but, for educational purposes, I would like to know which mqtt topics are used to create these sensors:

- total_usage
- firmware_update
- online
- print_bed_type
- print_weight
- print_lentgh
- start_time
- end_time

My printer is the A1.

Thanks in advance for your help!

**Comment by jjvelar** [comment]
Any thoughts on this, @AdrianGarside? Thanks!

**Comment by jjvelar** [comment]
Just wanted to share the template sensors I have created in Home Assistant to calculate print job start and end time.
I'd appreciate any comments or suggestions to improve the code as I am not very good on this :-)
In any case, I hope it can help anyone.

    #############################################
    # Bambu Lab A1 Print Job Start and End Time #
    #############################################
    - name: "Bambu Lab A1 Print Job Start Time"
      unique_id: bambu_lab_a1_print_job_start_time_01
      icon: mdi:clock-start
      state: >-
        {% set print_status = states('sensor.bambu_lab_a1_print_job_status') %}
        {% set start_time = states('sensor.bambu_lab_a1_print_job_start_time') %}
        {% if print_status == "Prepare" %}
            {% set start_time = now().strftime('%Y-%m-%d %H:%M:%S') %}
        {% else %}
            {% if print_status == "Slicing" or print_status == "Running" or print_status == "Pause" %}
                {% set start_time = start_time %}
            {% else %}
                {% set start_time = "" %}
            {% endif %}
        {% endif %}
        {{ start_time }}
      
    - name: "Bambu Lab A1 Print Job End Time"
      unique_id: bambu_lab_a1_print_job_end_time_01
      icon: mdi:clock-end
      state: >-
        {% set remaining_time = states('sensor.bambu_lab_a1_print_job_remaining_time')|int(0) %}
        {% if remaining_time > 0 %}
            {% set end_time = (now() + timedelta(minutes = remaining_time)).strftime('%Y-%m-%d %H:%M:%S') %}
        {% else %}
            {% set end_time = "" %}
        {% endif %}
        {{ end_time }}

**Comment by jjvelar** [comment]
Any thoughts on how to get total_usage, firmware_update, online and print_bed_type sensors?
Had a look into the repository but it is Klingon to me :-)
Thanks!

**Comment by AdrianGarside** [comment]
total_usage is calculated by the integration
firmware update is in the mqtt payload from the printer
print_bed_type is gotten from the bambu cloud API so is only available for cloud prints

**Reply by jjvelar** [reply]
Thanks!

**Comment by AdrianGarside** [comment]
In fact all that you listed except firmware update are either generated by the integration or gotten from the cloud API.
